TransactionalModePreservesAllData test is flaky

Unrelated failure observed in #13320.

```
Shouldly.ShouldAssertException : buildResult
should be
True
but was
False

   at Microsoft.Build.Tasks.UnitTests.WriteLinesToFile_Tests.TransactionalModePreservesAllData() in D:\a\_work\1\s\src\Tasks.UnitTests\WriteLinesToFile_Tests.cs:line 505
   at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```


```
C:\Users\cloudtest\AppData\Local\Temp\5rqrldid.v3q\MSBuildTempn4a553ru.i1z\Temporary3ec655280eb148a09a39ab8ab5eca76a\TestProject3.csproj(7,21):  error MSB3491: Could not write lines to file "C:\Users\cloudtest\AppData\Local\Temp\5rqrldid.v3q\MSBuildTempn4a553ru.i1z\Temporary3ec655280eb148a09a39ab8ab5eca76a\output.txt". Cannot create a file when that file already exists. 
Done executing task "WriteLinesToFile" -- FAILED.
Done building target "WriteToFile" in project "TestProject3.csproj" -- FAILED.
```

--- a/src/Tasks.UnitTests/WriteLinesToFile_Tests.cs
+++ b/src/Tasks.UnitTests/WriteLinesToFile_Tests.cs
@@ -453,7 +453,7 @@ namespace Microsoft.Build.Tasks.UnitTests
         }
 
         [Fact]
-        public void TransactionalModePreservesAllData()
+        public void TransactionalModeSucceedsWithConcurrentOverwrites()
         {
             using (var testEnv = TestEnvironment.Create(_output))
             {

--- a/src/Tasks/FileIO/WriteLinesToFile.cs
+++ b/src/Tasks/FileIO/WriteLinesToFile.cs
@@ -220,29 +220,38 @@ namespace Microsoft.Build.Tasks
                         // Only retry with Replace if the destination now exists (concurrent write race).
                         if (System.IO.File.Exists(filePath))
                         {
-                            IOException lastReplaceException = null;
-
                             // Retry Replace a few times with a small delay, mirroring the later Replace retry logic.
-                            for (int retry = 0; retry < 3; retry++)
+                            bool TryReplaceWithRetry(string sourcePath, string destinationPath, out IOException lastReplaceException)
                             {
-                                try
-                                {
-                                    if (retry > 0)
-                                    {
-                                        System.Threading.Thread.Sleep(10);
-                                    }
+                                lastReplaceException = null;
 
-                                    System.IO.File.Replace(temporaryFilePath, filePath, null, true);
-                                    temporaryFilePath = null; // Mark as successfully replaced
-                                    return !Log.HasLoggedErrors;
-                                }
-                                catch (IOException replaceEx)
+                                for (int retry = 0; retry < 3; retry++)
                                 {
-                                    lastReplaceException = replaceEx;
-                                    // Continue to next retry.
+                                    try
+                                    {
+                                        if (retry > 0)
+                                        {
+                                            System.Threading.Thread.Sleep(10);
+                                        }
+
+                                        System.IO.File.Replace(sourcePath, destinationPath, null, true);
+                                        return true;
+                                    }
+                                    catch (IOException replaceEx)
+                                    {
+                                        lastReplaceException = replaceEx;
+                                        // Continue to next retry.
+                                    }
                                 }
+
+                                return false;
                             }
 
+                            if (TryReplaceWithRetry(temporaryFilePath, filePath, out IOException lastReplaceException))
+                            {
+                                temporaryFilePath = null; // Mark as successfully replaced
+                                return !Log.HasLoggedErrors;
+                            }
                             // All Replace retries failed; log the original move error as the root cause.
                             string lockedFileMessage = LockCheck.GetLockedFileMessage(filePath);
                             string replaceMessage = lastReplaceException != null ? lastReplaceException.Message : string.Empty;

--- a/src/Tasks/FileIO/WriteLinesToFile.cs
+++ b/src/Tasks/FileIO/WriteLinesToFile.cs
@@ -220,47 +220,19 @@ namespace Microsoft.Build.Tasks
                         // Only retry with Replace if the destination now exists (concurrent write race).
                         if (System.IO.File.Exists(filePath))
                         {
-                            // Retry Replace a few times with a small delay, mirroring the later Replace retry logic.
-                            bool TryReplaceWithRetry(string sourcePath, string destinationPath, out IOException lastReplaceException)
+                            try
                             {
-                                lastReplaceException = null;
-
-                                for (int retry = 0; retry < 3; retry++)
-                                {
-                                    try
-                                    {
-                                        if (retry > 0)
-                                        {
-                                            System.Threading.Thread.Sleep(10);
-                                        }
-
-                                        System.IO.File.Replace(sourcePath, destinationPath, null, true);
-                                        return true;
-                                    }
-                                    catch (IOException replaceEx)
-                                    {
-                                        lastReplaceException = replaceEx;
-                                        // Continue to next retry.
-                                    }
-                                }
-
-                                return false;
-                            }
-
-                            if (TryReplaceWithRetry(temporaryFilePath, filePath, out IOException lastReplaceException))
-                            {
+                                System.IO.File.Replace(temporaryFilePath, filePath, null, true);
                                 temporaryFilePath = null; // Mark as successfully replaced
                                 return !Log.HasLoggedErrors;
                             }
-                            // All Replace retries failed; log the original move error as the root cause.
-                            string lockedFileMessage = LockCheck.GetLockedFileMessage(filePath);
-                            string replaceMessage = lastReplaceException != null ? lastReplaceException.Message : string.Empty;
-                            Log.LogErrorWithCodeFromResources(
-                                "WriteLinesToFile.ErrorOrWarning",
-                                filePath.OriginalValue,
-                                moveEx.Message + (string.IsNullOrEmpty(replaceMessage) ? string.Empty : " " + replaceMessage),
-                                lockedFileMessage);
-                            return !Log.HasLoggedErrors;
+                            catch (IOException replaceEx)
+                            {
+                                // Both attempts failed; log the original move error as the root cause.
+                                string lockedFileMessage = LockCheck.GetLockedFileMessage(filePath);
+                                Log.LogErrorWithCodeFromResources("WriteLinesToFile.ErrorOrWarning", filePath.OriginalValue, moveEx.Message + " " + replaceEx.Message, lockedFileMessage);
+                                return !Log.HasLoggedErrors;
+                            }
                         }
 
                         // Destination doesn't exist; move failed for a different reason.

--- a/src/Tasks/FileIO/WriteLinesToFile.cs
+++ b/src/Tasks/FileIO/WriteLinesToFile.cs
@@ -220,19 +220,38 @@ namespace Microsoft.Build.Tasks
                         // Only retry with Replace if the destination now exists (concurrent write race).
                         if (System.IO.File.Exists(filePath))
                         {
-                            try
+                            IOException lastReplaceException = null;
+
+                            // Retry Replace a few times with a small delay, mirroring the later Replace retry logic.
+                            for (int retry = 0; retry < 3; retry++)
                             {
-                                System.IO.File.Replace(temporaryFilePath, filePath, null, true);
-                                temporaryFilePath = null; // Mark as successfully replaced
-                                return !Log.HasLoggedErrors;
+                                try
+                                {
+                                    if (retry > 0)
+                                    {
+                                        System.Threading.Thread.Sleep(10);
+                                    }
+
+                                    System.IO.File.Replace(temporaryFilePath, filePath, null, true);
+                                    temporaryFilePath = null; // Mark as successfully replaced
+                                    return !Log.HasLoggedErrors;
+                                }
+                                catch (IOException replaceEx)
+                                {
+                                    lastReplaceException = replaceEx;
+                                    // Continue to next retry.
+                                }
                             }
-                            catch (IOException replaceEx)
-                            {
-                                // Both attempts failed; log the original move error as the root cause.
-                                string lockedFileMessage = LockCheck.GetLockedFileMessage(filePath);
-                                Log.LogErrorWithCodeFromResources("WriteLinesToFile.ErrorOrWarning", filePath.OriginalValue, moveEx.Message + " " + replaceEx.Message, lockedFileMessage);
-                                return !Log.HasLoggedErrors;
-                            }
+
+                            // All Replace retries failed; log the original move error as the root cause.
+                            string lockedFileMessage = LockCheck.GetLockedFileMessage(filePath);
+                            string replaceMessage = lastReplaceException != null ? lastReplaceException.Message : string.Empty;
+                            Log.LogErrorWithCodeFromResources(
+                                "WriteLinesToFile.ErrorOrWarning",
+                                filePath.OriginalValue,
+                                moveEx.Message + (string.IsNullOrEmpty(replaceMessage) ? string.Empty : " " + replaceMessage),
+                                lockedFileMessage);
+                            return !Log.HasLoggedErrors;
                         }
 
                         // Destination doesn't exist; move failed for a different reason.

--- a/src/Tasks/FileIO/WriteLinesToFile.cs
+++ b/src/Tasks/FileIO/WriteLinesToFile.cs
@@ -215,23 +215,30 @@ namespace Microsoft.Build.Tasks
                         temporaryFilePath = null; // Mark as successfully moved
                         return !Log.HasLoggedErrors;
                     }
-                    catch (IOException)
+                    catch (IOException moveEx)
                     {
-                        // File may have been created by another thread between Replace and Move.
-                        // Retry with Replace since the file now exists.
-                        try
+                        // Only retry with Replace if the destination now exists (concurrent write race).
+                        if (System.IO.File.Exists(filePath))
                         {
-                            System.IO.File.Replace(temporaryFilePath, filePath, null, true);
-                            temporaryFilePath = null; // Mark as successfully replaced
-                            return !Log.HasLoggedErrors;
+                            try
+                            {
+                                System.IO.File.Replace(temporaryFilePath, filePath, null, true);
+                                temporaryFilePath = null; // Mark as successfully replaced
+                                return !Log.HasLoggedErrors;
+                            }
+                            catch (IOException replaceEx)
+                            {
+                                // Both attempts failed; log the original move error as the root cause.
+                                string lockedFileMessage = LockCheck.GetLockedFileMessage(filePath);
+                                Log.LogErrorWithCodeFromResources("WriteLinesToFile.ErrorOrWarning", filePath.OriginalValue, moveEx.Message + " " + replaceEx.Message, lockedFileMessage);
+                                return !Log.HasLoggedErrors;
+                            }
                         }
-                        catch (IOException moveEx)
-                        {
-                            // Move failed, log and return
-                            string lockedFileMessage = LockCheck.GetLockedFileMessage(filePath);
-                            Log.LogErrorWithCodeFromResources("WriteLinesToFile.ErrorOrWarning", filePath.OriginalValue, moveEx.Message, lockedFileMessage);
-                            return !Log.HasLoggedErrors;
-                        }
+
+                        // Destination doesn't exist; move failed for a different reason.
+                        string lockedFileDiagnostics = LockCheck.GetLockedFileMessage(filePath);
+                        Log.LogErrorWithCodeFromResources("WriteLinesToFile.ErrorOrWarning", filePath.OriginalValue, moveEx.Message, lockedFileDiagnostics);
+                        return !Log.HasLoggedErrors;
                     }
                 }
                 catch (IOException)

--- a/src/Tasks/FileIO/WriteLinesToFile.cs
+++ b/src/Tasks/FileIO/WriteLinesToFile.cs
@@ -225,11 +225,11 @@ namespace Microsoft.Build.Tasks
                             temporaryFilePath = null; // Mark as successfully replaced
                             return !Log.HasLoggedErrors;
                         }
-                        catch (IOException moveEx)
+                        catch (IOException replaceEx)
                         {
-                            // Move failed, log and return
+                            // Replace failed, log and return
                             string lockedFileMessage = LockCheck.GetLockedFileMessage(filePath);
-                            Log.LogErrorWithCodeFromResources("WriteLinesToFile.ErrorOrWarning", filePath.OriginalValue, moveEx.Message, lockedFileMessage);
+                            Log.LogErrorWithCodeFromResources("WriteLinesToFile.ErrorOrWarning", filePath.OriginalValue, replaceEx.Message, lockedFileMessage);
                             return !Log.HasLoggedErrors;
                         }
                     }

--- a/src/Tasks/FileIO/WriteLinesToFile.cs
+++ b/src/Tasks/FileIO/WriteLinesToFile.cs
@@ -22,6 +22,29 @@ namespace Microsoft.Build.Tasks
         // Default encoding taken from System.IO.WriteAllText()
         private static readonly Encoding s_defaultEncoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
 
+        /// <summary>
+        /// Moves <paramref name="source"/> to <paramref name="destination"/>, overwriting it if it exists.
+        /// Uses <c>Microsoft.IO.File</c> on .NET Framework to access the overwrite overload,
+        /// which is available natively in <c>System.IO.File</c> on .NET 6+.
+        /// </summary>
+        private static void MoveFileWithOverwrite(string source, string destination)
+        {
+#if NETFRAMEWORK
+            // Microsoft.IO.Redist backports File.Move(overwrite) to .NET Framework.
+            Microsoft.IO.File.Move(source, destination, overwrite: true);
+#elif NET
+            // File.Move(overwrite) is available natively on .NET 5+.
+            System.IO.File.Move(source, destination, overwrite: true);
+#else
+            // netstandard2.0: overwrite overload unavailable; delete destination first.
+            if (System.IO.File.Exists(destination))
+            {
+                System.IO.File.Delete(destination);
+            }
+            System.IO.File.Move(source, destination);
+#endif
+        }
+
         /// <inheritdoc />
         public TaskEnvironment TaskEnvironment { get; set; } = TaskEnvironment.Fallback;
 
@@ -181,8 +204,10 @@ namespace Microsoft.Build.Tasks
         }
 
         /// <summary>
-        /// Saves content to file atomically using a temporary file, following the Visual Studio editor pattern.
-        /// This is for overwrite mode where we write the entire content.
+        /// Saves content to file atomically using a temporary file.
+        /// Writes to a temp file first, then moves it to the target with overwrite,
+        /// which handles both the "target exists" and "target doesn't exist" cases in a
+        /// single call and eliminates the race window present in a check-then-act pattern.
         /// </summary>
         private bool SaveAtomically(AbsolutePath filePath, string contentsAsString, Encoding encoding)
         {
@@ -197,81 +222,33 @@ namespace Microsoft.Build.Tasks
                 // Write content to temporary file
                 System.IO.File.WriteAllText(temporaryFilePath, contentsAsString, encoding);
 
-                // Attempt to atomically replace target file with temporary file
-                try
+                // Atomically move temp file to target, overwriting if it already exists.
+                // Using overwrite: true handles concurrent writes without a race condition —
+                // both "target doesn't exist" and "target already exists" cases are covered
+                // by a single operation, with no window between them.
+                const int maxAttempts = 3;
+                Exception lastException = null;
+                for (int attempt = 1; attempt <= maxAttempts; attempt++)
                 {
-                    // Replace the contents of filePath with the contents of the temporary using File.Replace
-                    // to preserve the various attributes of the original file.
-                    System.IO.File.Replace(temporaryFilePath, filePath, null, true);
-                    temporaryFilePath = null; // Mark as successfully replaced
-                    return !Log.HasLoggedErrors;
-                }
-                catch (FileNotFoundException)
-                {
-                    // The target file doesn't exist, which is fine. Move the temp file to target.
-                    try
+                    if (attempt > 1)
                     {
-                        System.IO.File.Move(temporaryFilePath, filePath);
-                        temporaryFilePath = null; // Mark as successfully moved
-                        return !Log.HasLoggedErrors;
-                    }
-                    catch (IOException moveEx)
-                    {
-                        // Only retry with Replace if the destination now exists (concurrent write race).
-                        if (System.IO.File.Exists(filePath))
-                        {
-                            try
-                            {
-                                System.IO.File.Replace(temporaryFilePath, filePath, null, true);
-                                temporaryFilePath = null; // Mark as successfully replaced
-                                return !Log.HasLoggedErrors;
-                            }
-                            catch (IOException replaceEx)
-                            {
-                                // Both attempts failed; log the original move error as the root cause.
-                                string lockedFileMessage = LockCheck.GetLockedFileMessage(filePath);
-                                Log.LogErrorWithCodeFromResources("WriteLinesToFile.ErrorOrWarning", filePath.OriginalValue, moveEx.Message + " " + replaceEx.Message, lockedFileMessage);
-                                return !Log.HasLoggedErrors;
-                            }
-                        }
-
-                        // Destination doesn't exist; move failed for a different reason.
-                        string lockedFileDiagnostics = LockCheck.GetLockedFileMessage(filePath);
-                        Log.LogErrorWithCodeFromResources("WriteLinesToFile.ErrorOrWarning", filePath.OriginalValue, moveEx.Message, lockedFileDiagnostics);
-                        return !Log.HasLoggedErrors;
-                    }
-                }
-                catch (IOException)
-                {
-                    // Replace failed (likely file is locked). Retry a few times with small delay.
-                    for (int retry = 1; retry < 3; retry++)
-                    {
-                        try
-                        {
-                            System.Threading.Thread.Sleep(10);
-                            System.IO.File.Replace(temporaryFilePath, filePath, null, true);
-                            temporaryFilePath = null; // Mark as successfully replaced
-                            return !Log.HasLoggedErrors;
-                        }
-                        catch (IOException)
-                        {
-                            // Continue to next retry
-                        }
+                        // Log the retry so concurrency issues are visible when diagnosing builds.
+                        Log.LogMessageFromResources(MessageImportance.High, "WriteLinesToFile.Retry",
+                            filePath.OriginalValue, attempt, maxAttempts, lastException.Message);
+                        System.Threading.Thread.Sleep(10);
                     }
 
-                    // Retries exhausted. Try simple write as fallback.
                     try
                     {
-                        System.IO.File.WriteAllText(filePath, contentsAsString, encoding);
-                        temporaryFilePath = null; // Mark temp as not needed
+                        MoveFileWithOverwrite(temporaryFilePath, filePath);
+                        temporaryFilePath = null;
                         return !Log.HasLoggedErrors;
                     }
-                    catch (Exception fallbackEx) when (ExceptionHandling.IsIoRelatedException(fallbackEx))
+                    catch (Exception ex) when (attempt < maxAttempts && ExceptionHandling.IsIoRelatedException(ex))
                     {
-                        string lockedFileMessage = LockCheck.GetLockedFileMessage(filePath);
-                        Log.LogErrorWithCodeFromResources("WriteLinesToFile.ErrorOrWarning", filePath.OriginalValue, fallbackEx.Message, lockedFileMessage);
-                        return !Log.HasLoggedErrors;
+                        lastException = ex;
                     }
+                    // On the last attempt, the exception propagates to the outer handler.
                 }
             }
             catch (Exception e) when (ExceptionHandling.IsIoRelatedException(e))
@@ -298,6 +275,8 @@ namespace Microsoft.Build.Tasks
                     }
                 }
             }
+
+            return !Log.HasLoggedErrors;
         }
 
         /// <summary>

--- a/src/Tasks/FileIO/WriteLinesToFile.cs
+++ b/src/Tasks/FileIO/WriteLinesToFile.cs
@@ -215,12 +215,23 @@ namespace Microsoft.Build.Tasks
                         temporaryFilePath = null; // Mark as successfully moved
                         return !Log.HasLoggedErrors;
                     }
-                    catch (IOException moveEx)
+                    catch (IOException)
                     {
-                        // Move failed, log and return
-                        string lockedFileMessage = LockCheck.GetLockedFileMessage(filePath);
-                        Log.LogErrorWithCodeFromResources("WriteLinesToFile.ErrorOrWarning", filePath.OriginalValue, moveEx.Message, lockedFileMessage);
-                        return !Log.HasLoggedErrors;
+                        // File may have been created by another thread between Replace and Move.
+                        // Retry with Replace since the file now exists.
+                        try
+                        {
+                            System.IO.File.Replace(temporaryFilePath, filePath, null, true);
+                            temporaryFilePath = null; // Mark as successfully replaced
+                            return !Log.HasLoggedErrors;
+                        }
+                        catch (IOException moveEx)
+                        {
+                            // Move failed, log and return
+                            string lockedFileMessage = LockCheck.GetLockedFileMessage(filePath);
+                            Log.LogErrorWithCodeFromResources("WriteLinesToFile.ErrorOrWarning", filePath.OriginalValue, moveEx.Message, lockedFileMessage);
+                            return !Log.HasLoggedErrors;
+                        }
                     }
                 }
                 catch (IOException)

--- a/src/Tasks/FileIO/WriteLinesToFile.cs
+++ b/src/Tasks/FileIO/WriteLinesToFile.cs
@@ -36,12 +36,8 @@ namespace Microsoft.Build.Tasks
             // File.Move(overwrite) is available natively on .NET 5+.
             System.IO.File.Move(source, destination, overwrite: true);
 #else
-            // netstandard2.0: overwrite overload unavailable; delete destination first.
-            if (System.IO.File.Exists(destination))
-            {
-                System.IO.File.Delete(destination);
-            }
-            System.IO.File.Move(source, destination);
+            // netstandard2.0 output is ref asm only and never executed at runtime.
+            throw new PlatformNotSupportedException();
 #endif
         }
 

